nifi-nar-bundles/nifi-azure-bundle/nifi-azure-graph-authorizer/src/main/java/org/apache/nifi/authorization/azure/AzureGraphUserGroupProvider.java

```diff
@@ -0,0 +1,420 @@
+            throw new AuthorizerCreationException(String.format("%s is a required field for AzureGraphUserGroupProvder",
+                    APP_REG_CLIENT_SECRET_PROPERTY));
+        }
+        if (this.pageSize > 999) {
+            throw new AuthorizerCreationException("Max page size for graph rest api call is 999.");
+        }
+
+        try {
+            authProvider = new ClientCredentialAuthProvider.Builder()
+                .authorityEndpoint(authorityEndpoint)
+                .tenantId(tenantId)
+                .clientId(clientId)
+                .clientSecret(clientSecret)
+                .build();
+            graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();
+        } catch (ClientException ep) {
+            throw new AuthorizerCreationException("Failed to create a GraphServiceClient", ep);
+        }
+
+        // first, load list of group name if there is any prefix, suffix, substring
+        // filter defined, paging thru groups.
+        // then, add additonal group list if there is group list inclusion defined.
+        final String prefix = getProperty(configurationContext, GROUP_FILTER_PREFIX_PROPERTY, null);
+        final String suffix = getProperty(configurationContext, GROUP_FILTER_SUFFIX_PROPERTY, null);
+        final String substring = getProperty(configurationContext, GROUP_FILTER_SUBSTRING_PROPERTY, null);
+        final String groupFilterLst = getProperty(configurationContext, GROUP_FILTER_LIST_PROPERTY, null);
+
+        // if no group filter is specified, generate exception since we don't want to
+        // load whole groups from AAD.
+        if (StringUtils.isBlank(prefix) && StringUtils.isBlank(suffix) && StringUtils.isBlank(substring)
+                && StringUtils.isBlank(groupFilterLst)) {
+            throw new AuthorizerCreationException("At least one GROUP_FILTER should be specified");
+        }
+        if (!StringUtils.isBlank(prefix) || !StringUtils.isBlank(suffix) || !StringUtils.isBlank(substring)) {
+            this.groupFilterSet = getGroupListWith(prefix, suffix, substring, pageSize);
+        }
+
+        if (!StringUtils.isBlank(groupFilterLst)) {
+            final Set<String> groupFilterNames = Arrays.stream(groupFilterLst.split(",")).map(String::trim)
+                    .collect(Collectors.toSet());
+            if (groupFilterSet != null) {
+                Set<String> gNameSet = this.groupFilterSet.stream().collect(Collectors.toSet());
+                this.groupFilterSet
+                        .addAll(groupFilterNames.stream().filter(gname -> !gNameSet.contains(gname)).collect(Collectors.toSet()));
+
+            } else {
+                groupFilterSet = groupFilterNames;
+            }
+        }
+        try {
+            refreshUserGroupData(groupFilterSet);
+        } catch (IOException | ClientException ep) {
+            throw new AuthorizerCreationException("Failed to load user/group from Azure AD", ep);
+        }
```

Comment:
As far as I can tell the `groupFilterSet` never gets refreshed so group additions and removals aren't taking affect. My suggestion would be that we refactor this so `refreshUserGroupData()` includes the `getGroupList()` logic. See https://gist.github.com/jfrazee/2039586480267717cba4b0ac35e736ae for an example of what I'm thinking (you can ignore the additional debugging).